src/CodeGen_ARM.cpp

```diff
@@ -360,9 +374,57 @@ void CodeGen_ARM::visit(const Mul *op) {
         return;
     }
 
+    Type t = op->type;
+    vector<Expr> matches;
+
+    int shift_amount = 0;
+    if (is_const_power_of_two_integer(op->b, &shift_amount)) {
+        // Let LLVM handle these.
+        CodeGen_Posix::visit(op);
+        return;
+    }
+
+    // LLVM really struggles to generate mlal unless we generate mull intrinsics
+    // for the multiplication part first.
+    for (size_t i = 0; i < multiplies.size() ; i++) {
+        const Pattern &pattern = multiplies[i];
+        //debug(4) << "Trying pattern: " << patterns[i].intrin << " " << patterns[i].pattern << "\n";
+        if (expr_match(pattern.pattern, op, matches)) {
+
+            //debug(4) << "Match!\n";
+            if (pattern.type == Pattern::Simple) {
+                value = call_pattern(pattern, t, matches);
+                return;
+            } else if (pattern.type == Pattern::NarrowArgs) {
+                Type narrow_t = t.with_bits(t.bits() / 2);
+                // Try to narrow all of the args.
+                bool all_narrow = true;
+                for (size_t i = 0; i < matches.size(); i++) {
+                    internal_assert(matches[i].type().bits() == t.bits());
+                    internal_assert(matches[i].type().lanes() == t.lanes());
+                    // debug(4) << "Attemping to narrow " << matches[i] << " to " << t << "\n";
+                    matches[i] = lossless_cast(narrow_t, matches[i]);
+                    if (!matches[i].defined()) {
+                        // debug(4) << "failed\n";
+                        all_narrow = false;
+                    } else {
+                        // debug(4) << "success: " << matches[i] << "\n";
+                        internal_assert(matches[i].type() == narrow_t);
+                    }
+                }
+
+                if (all_narrow) {
+                    value = call_pattern(pattern, t, matches);
+                    return;
+                }
+            }
+        }
+    }
+
     // Vector multiplies by 3, 5, 7, 9 should do shift-and-add or
     // shift-and-sub instead to reduce register pressure (the
     // shift is an immediate)
+    // TODO: Verify this is still good codegen.
     if (is_const(op->b, 3)) {
         value = codegen(op->a*2 + op->a);
```

Comment:
This is dubious because it codegens op->a twice. Consider the horrific bitcode we'd produce with ((((((x*3 + y)*3 + z)*3 + w)*3 + u)*3). Might be better to just delete this. Probably a task for a separate PR unless you've got nothing better to do on a Saturday morning.

Reply:
Let's do it in a separate PR. Since both changes will have a pretty wide ranging impact on ARM codegen, I'd like to make them separately with some soak time between them.